firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -0,0 +1,358 @@
+  private var receiveChannel: Channel<Frame> = Channel()
+  private var functionCallChannel: Channel<List<FunctionCallPart>> = Channel()
+
+  internal class ClientContentSetup(val turns: List<Content.Internal>, val turnComplete: Boolean) {
+    @Serializable
+    internal class Internal(@SerialName("client_content") val clientContent: ClientContent) {
+      @Serializable
+      internal data class ClientContent(
+        @SerialName("turns") val turns: List<Content.Internal>,
+        @SerialName("turn_complete") val turnComplete: Boolean
+      )
+    }
+
+    fun toInternal() = Internal(Internal.ClientContent(turns, turnComplete))
+  }
+
+  internal class ToolResponseSetup(
+    val functionResponses: List<FunctionResponsePart.Internal.FunctionResponse>
+  ) {
+
+    @Serializable
+    internal data class Internal(val toolResponse: ToolResponse) {
+      @Serializable
+      internal data class ToolResponse(
+        val functionResponses: List<FunctionResponsePart.Internal.FunctionResponse>
+      )
+    }
+
+    fun toInternal() = Internal(Internal.ToolResponse(functionResponses))
+  }
+
+  internal class ServerContentSetup(val modelTurn: Content.Internal) {
+    @Serializable
+    internal class Internal(@SerialName("serverContent") val serverContent: ServerContent) {
+      @Serializable
+      internal data class ServerContent(@SerialName("modelTurn") val modelTurn: Content.Internal)
+    }
+
+    fun toInternal() = Internal(Internal.ServerContent(modelTurn))
+  }
+
+  internal class MediaStreamingSetup(val mediaChunks: List<MediaData.Internal>) {
+    @Serializable
+    internal class Internal(val realtimeInput: MediaChunks) {
+      @Serializable internal data class MediaChunks(val mediaChunks: List<MediaData.Internal>)
+    }
+    fun toInternal() = Internal(Internal.MediaChunks(mediaChunks))
+  }
+
+  internal data class ToolCallSetup(
+    val functionCalls: List<FunctionCallPart.Internal.FunctionCall>
+  ) {
+
+    @Serializable
+    internal class Internal(val toolCall: ToolCall) {
+
+      @Serializable
+      internal data class ToolCall(val functionCalls: List<FunctionCallPart.Internal.FunctionCall>)
+    }
+
+    fun toInternal(): Internal {
+      return Internal(Internal.ToolCall(functionCalls))
+    }
+  }
+
+  /**
+   * Receives all function call responses from the server for the audio conversation feature.
+   *
+   * @return A [Flow] which will emit list of [FunctionCallPart] as they are returned by the model.
+   */
+  public fun receiveAudioConversationFunctionCalls(): Flow<List<FunctionCallPart>> {
+    return functionCallChannel.receiveAsFlow()
+  }
+
+  private fun fillRecordedAudioQueue() {
+    CoroutineScope(Dispatchers.IO).launch {
+      audioHelper!!.startRecording().collect {
+        if (!isRecording) {
+          cancel()
+        }
+        audioQueue.add(it)
+      }
+    }
+  }
+
+  private fun sendAudioDataToServer() {
+    CoroutineScope(Dispatchers.Default).launch {
```

Comment:
Why are we specifying the scope? Ideally we just inherit the scope as a suspend function (like we do for the standard API)

Reply:
we would want to run different operations in different threads.. We are doing a lot of the heavy lifting for the developer in the startAudioConversation() similar to how a developer would implement the audioConversation